# Minion Bioinformatics Workshop

Informal workshop led by Oliver Stringham presented to Julie Lockwood's eDNA lab (~5 people) about a bioinformatics pipeline for sequence data from a MinION sequencer. Oliver is giving weekly workshop presentation with accompanying code and we go through each step together. Oliver has made this repository public so that other people can access the material and learn from it.

For Weeks 3 through 6, the raw fastq data used can be found here: [https://github.com/ocstringham/minion-bioinformatics-workshop-data](https://github.com/ocstringham/minion-bioinformatics-workshop-data) under the fastq folder.

## Topics Covered

* Week 1: Bioinformatics Overview
* Week 2: Software Overview
* Week 3: Concatenating Output FASTQ files, Quality Filtering, Length Filter, Trimming Primers
* Week 4: Trimming ONT & Clustering Sequences to MOTUs
* Week 5: Reference Database to BLAST
* Week 6: Clean BLAST & compile results
* Week 7: Reference database creation overview

## Notes on other topics to potentially cover

* Basecalling and demultiplexing in command line using Dorado
* Github
* Make a reference database the crab/long/more comprehensive way (ie not just using Midori) [do next]
* Make a bash function
* Make an r script a command line function
* processing contaminants, negative controls, and positive controls

